lib/mongo/server/connection.rb

```diff
@@ -155,7 +155,15 @@ def ping
 
       def deliver(messages)
         write(messages)
-        messages.last.replyable? ? read : nil
+
+        if messages.last.replyable?
+          # Protection against returning the response to a previous request. See
+          # RUBY-1117
+          reply = read until reply && reply.response_to == messages.last.request_id
```

Comment:
I see what you're saying but if the application is killing threads in a number of places after initiating long-running operations, it should probably change the socket_timeout option to better match the permitted execution time of threads or use the max_time_ms option. You could also close the client if it's a real threat that future operations would read responses from previous ones.

Perhaps this isn't clear in the code but a Connection object is a wrapper around a Socket. In the #with_connection method, the connection does not have to be disconnected when there is an error because the socket is disconnected at a lower level. If you look at the #ensure_connected method in Connectable, the socket is disconnected when there is an exception so that it won't be used again:
https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/server/connectable.rb#L90-L95

I don't know much about Thread.handle_interrupt but will look into it. Thanks!

Feel free to add the op_code check (it should be pretty simple to add) or I can add it in separately later. Up to you and Vincent = )